### ISSUE TITLE
[社区任务5] 基于megatron+vllm_ascend完成qwen2.5 1.5b的GDPO训练

### ISSUE THREAD

**Issue by wucong25** [verl-project/verl-ascend-recipe#18]
- 任务描述
"昇腾上基于VeRL开放仓库的vllm_ascend推理后端，补齐该模型 GDPO训练对于megatron后端的支持。
本期任务的具体信息如下：
（1）模型名称与算法：qwen2.5 1.5b + GDPO
（2）VeRL仓库的模型源码链接：
https://github.com/verl-project/verl
（3）参考脚本：examples\gdpo_trainer\run_qwen1_5b_gdpo.sh
（4） 开发平台：Atlas 800T A2或者A3"

- 验收标准
一、任务交付件
""一、任务交付件
本期任务为基于verl开放仓库代码进行算法扩展，请合入开发代码。主要开发点如下：
（1） 模型适配patch
（2） 算法优化实现
二、验收标准：
1）精度/性能要求
精度：
长跑12小时或者训练100步，reward上升，平均误差小于5%，下游评测误差5%内。如果无GPU标杆，reward上升即可。
性能：
A2上达成A100的0.7倍以上。如果无GPU标杆，TPS绝对值大于100即可。
2）实践文档：
算法适配调优文档1篇
3）任务完成标准
本次任务完成标准为：
精度/性能（根据实际要求）达标，PR完成合入，实践文档提交到仓库issue。""

- 任务类型
算法支持

**Comment by swallow-design** [comment]
认领这个任务

**Comment by Ronald1995** [comment]
> 认领这个任务

@swallow-design 

欢迎认领任务，请参考https://github.com/verl-project/verl/issues/6439  社区任务池明确该任务的：

- 完成的截止日期
- 开发进展反馈
- 微信答疑群
- 任务交付注意事项

等信息。如果您同时认领了多项任务，但无法都能进行投入，可以在部分任务中回复退出.

麻烦您加入到对应微信群，群备注名修改为"社区任务+您的gitcode账号", 后续有相关消息和问题都可以在微信群咨询答疑。 等您加入到微信群后，我这边会在社区任务池里面登记任务责任人。

**Comment by jianzhnie** [comment]
认领这个任务

**Comment by chengminhua** [comment]
> 认领这个任务
你好，请关注下https://github.com/verl-project/verl/issues/6439 里的任务规则，每周一更新下进展哦

**Comment by chengminhua** [comment]
> 认领这个任务

欢迎参加认领社区任务，当前一个任务只分配一个开发，根据评论认领时间确定，您可以后续观察该任务是否被释放退出，或者认领其他任务 https://gitcode.com/org/Ascend/discussions/4

**Comment by chengminhua** [comment]
> 认领这个任务

你好，请关注下https://github.com/verl-project/verl/issues/6439 里的任务规则，每周一更新下进展哦

**Comment by chengminhua** [comment]
> 认领这个任务

感谢您对昇腾开源社区的关注与支持！
由于该任务在认领后较长时间内暂未收到进展更新，为保障项目整体推进节奏，我们将暂时结束本次认领，并重新开放该任务，供其他社区贡献者参与。
后续如您仍有兴趣参与昇腾开源项目，欢迎持续关注【昇腾生态官网】及本社区发布的最新任务，也欢迎通过提交 Issue或参与社区讨论等方式继续贡献。
再次感谢您的理解与支持，期待未来继续与您合作！

**Comment by RordChang** [comment]
认领这个任务

**Comment by chengminhua** [comment]
> 认领这个任务

感谢您此前对该任务的关注与支持！
由于上一位认领者暂未继续推进该任务，目前该任务已重新开放。根据此前的候选顺序，我们希望优先邀请您参与该任务。
如您目前仍有意愿认领，请在5个工作日内回复确认，并简要说明预计开始时间或初步推进计划。确认后，我们将为您保留任务，并提供必要的背景资料和支持。
如您近期时间安排不便，也请直接告知，我们将继续顺延联系其他候选贡献者，不会影响您后续参与其他社区任务。
再次感谢您对昇腾开源社区的关注，期待与您合作！

**Comment by RordChang** [comment]
> > 认领这个任务
> 
> 感谢您此前对该任务的关注与支持！ 由于上一位认领者暂未继续推进该任务，目前该任务已重新开放。根据此前的候选顺序，我们希望优先邀请您参与该任务。 如您目前仍有意愿认领，请在5个工作日内回复确认，并简要说明预计开始时间或初步推进计划。确认后，我们将为您保留任务，并提供必要的背景资料和支持。 如您近期时间安排不便，也请直接告知，我们将继续顺延联系其他候选贡献者，不会影响您后续参与其他社区任务。 再次感谢您对昇腾开源社区的关注，期待与您合作！

确认认领

**Comment by fangweii1** [comment]
> > > 认领这个任务
> > 
> > 
> > 感谢您此前对该任务的关注与支持！ 由于上一位认领者暂未继续推进该任务，目前该任务已重新开放。根据此前的候选顺序，我们希望优先邀请您参与该任务。 如您目前仍有意愿认领，请在5个工作日内回复确认，并简要说明预计开始时间或初步推进计划。确认后，我们将为您保留任务，并提供必要的背景资料和支持。 如您近期时间安排不便，也请直接告知，我们将继续顺延联系其他候选贡献者，不会影响您后续参与其他社区任务。 再次感谢您对昇腾开源社区的关注，期待与您合作！
> 
> 确认认领

你好，请关注下https://github.com/verl-project/verl/issues/6439 里的任务规则，每周一更新下进展哦

**Comment by RordChang** [comment]
已经完成任务训练，正在整理文档

**Comment by RordChang** [comment]
## GDPO 适配设计与复现结果

### 1. 任务目标

在 Ascend NPU 上完成 Qwen2.5-1.5B-Instruct 的 GDPO 训练，完整组合为：

- actor/reference：Megatron + MBridge
- rollout：vLLM-Ascend
- 算法：GDPO
- 设备：4 x Ascend 910B4
- 训练长度：100 steps

代码 PR：[verl-project/verl#7445](https://github.com/verl-project/verl/pull/7445)
Recipe仓 PR: https://github.com/verl-project/verl-ascend-recipe/pull/86
### 2. 实现设计

训练入口为：

```text
examples/gdpo_trainer/run_qwen2_5_1_5b_megatron.sh
```

主要调用链如下：

1. `model_engine=megatron` 启用 Megatron actor/reference，并通过 MBridge 完成模型转换。
2. rollout 使用 vLLM-Ascend，TP=2，默认每个 prompt 采样 4 个 responses。
3. `algorithm.adv_estimator=gdpo` 启用 GDPO advantage estimator。
4. RLLA scoring function 分别返回 `accuracy_reward` 和 `format_reward`。
5. GDPO 在每个奖励维度内独立进行 group normalization，再汇总为最终 advantage。

关键配置：

```text
algorithm.adv_estimator=gdpo
algorithm.gdpo_reward_keys=["accuracy_reward","format_reward"]
reward.reward_manager.name=gdpo
reward.custom_reward_function.path=verl/utils/reward_score/rlla.py
reward.custom_reward_function.name=compute_score
model_engine=megatron
```

本次同时补充了 CPU 测试，覆盖：

- 多奖励维度分别归一化后再求和
- 声明的奖励分量缺失时明确报错
- GDPO estimator/reward manager registry
- RLLA total reward 和分量 reward 输出

### 3. 数据准备

使用 `RamAnanth1/toolrl-rlla4k`，包含 3920 条训练数据和 80 条测试数据：

```bash
python - <<'PY'
from pathlib import Path
from datasets import load_dataset

root = Path("/path/to/data/rlla_4k")
root.mkdir(parents=True, exist_ok=True)

dataset = load_dataset("RamAnanth1/toolrl-rlla4k")
dataset["train"].to_parquet(root / "train.parquet")
dataset["test"].to_parquet(root / "test.parquet")
PY
```

最终数据布局：

```text
/path/to/data/rlla_4k/train.parquet
/path/to/data/rlla_4k/test.parquet
```

### 4. 运行命令

```bash
MODEL_PATH=/path/to/Qwen2.5-1.5B-Instruct \
DATA_ROOT=/path/to/data \
NPUS_PER_NODE=4 \
bash examples/gdpo_trainer/run_qwen2_5_1_5b_megatron.sh
```

脚本默认配置：

- train batch size：16
- PPO mini batch size：8
- prompt/response length：2048/1024
- Megatron actor/reference TP：2
- vLLM rollout TP：2
- rollout n：4
- total training steps：100
- resume mode：auto
- 按时间戳保存控制台日志

所有关键参数均可通过环境变量或附加 Hydra overrides 覆盖。

### 5. 验证环境

- 4 x Ascend 910B4
- CANN 9.0.0
- Python 3.11.15
- torch 2.9.0
- torch-npu 2.9.0.post2
- vLLM 0.18.0
- vLLM-Ascend 0.18.0
- Megatron-Core/MindSpeed 0.16.0
- MBridge 0.15.1

### 6. 训练结果

连续完成 100/100 steps，所有必需指标均可解析且为有限值。

| 指标 | 结果 |
|---|---:|
| 前 10 步 reward 均值 | -2.0765 |
| 后 10 步 reward 均值 | -0.3575 |
| Reward 变化 | +1.7189 |
| Reward 线性斜率 | +0.01421/step |
| 平均吞吐 | 427.1 token/s/NPU |
| 四卡合计吞吐 | 1708.4 token/s |
| 总 step time | 3419.12 s |

Reward 按窗口均值呈明确上升趋势，吞吐超过任务在无 GPU 标杆情况下的 TPS 要求。

### 7. 代码与检查

- PR：[verl-project/verl#7445](https://github.com/verl-project/verl/pull/7445)
- CPU tests：`4 passed, 1 warning`
- 完整 pre-commit：全部通过
- PR title checker：通过
- example naming checker：通过
- `bash -n`：通过

### 8. 完整训练日志

- [100-step 完整训练日志（已脱敏）](https://gist.github.com/RordChang/026f04feae655f3fca610890e41d9895)

当前训练、代码、测试和复现材料均已提交，等待 PR review 与合入。
